pyccel/parser/semantic.py

```diff
@@ -3785,212 +3911,263 @@ def _visit_FunctionDef(self, expr):
         for t,v in templates.items():
             templates[t] = UnionTypeAnnotation(*[self._visit(vi) for vi in v])
 
+        def unpack(ann):
+            if isinstance(ann, UnionTypeAnnotation):
+                return ann.type_list
+            else:
+                return [ann]
+
         # Filter out unused templates
-        templatable_args = [a.annotation for a in expr.arguments if isinstance(a.annotation, (SyntacticTypeAnnotation, UnionTypeAnnotation))]
-        arg_annotations = [annot for a in templatable_args for annot in (a.type_list \
-                                        if isinstance(a, UnionTypeAnnotation) else [a]) \
-                                        if isinstance(annot, SyntacticTypeAnnotation)]
-        used_type_names = set(a.dtype for a in arg_annotations)
+        templatable_args = [unpack(a.annotation) for a in expr.arguments if isinstance(a.annotation, (SyntacticTypeAnnotation, UnionTypeAnnotation))]
+        arg_annotations  = [annot for a in templatable_args for annot in a if isinstance(annot, SyntacticTypeAnnotation)]
+        used_type_names  = set(a.dtype for a in arg_annotations)
         templates = {t: v for t,v in templates.items() if t in used_type_names}
 
+        tmp_templates = {}
+        new_expr_args = []
+        for a in expr.arguments:
+            if isinstance(a.annotation, UnionTypeAnnotation):
+                annotation = [aa.type_list if isinstance(aa, UnionTypeAnnotation) else [aa] for aa in a.annotation]
+                annotation = [aa for a in annotation for aa in a]
+            else:
+                annotation = [a.annotation]
+            if len(annotation)>1:
+                tmp_template_name = a.name + '_' + random_string(12)
```

Comment:
here I want to create a random name for a temporary template, I don't want it to affect the printed code because it is just internally used.
I can do this instead `tmp_template_name = self.scope.get_new_name(tmp_template_name)`

Reply:
Is this name not preserved then? I thought it would be printed in the Python code.

> I can do this instead

What is the difference with my suggestion?

Reply:
So that it won't affect the generation of a new temp variable with the same name as `a`,
otherwise the counter in `create_incremented_string` will be different. but in any case it is not that important lol.

Reply:
> So that it won't affect the generation of a new temp variable with the same name as a,

Why would it affect the generation of a new temp variable with the same name as a? `a` will be added to the symbols in the syntactic stage so in any case the name that is generated here will be something like `a_00000` so as not to conflict with the existing symbol.

> but in any case it is not that important lol.

Any associated bug will not appear very often and can probably be avoided by rerunning but it is still a possible source of bugs (if a global variable has the same name as the template type). I would rather fix it while we remember the cause

Reply:
yes, that is why I suggested `mp_template_name = self.scope.get_new_name(tmp_template_name)` because you are correct, we might have collisions even if the probability is low

Reply:
The difference between:
`tmp_template_name = self.scope.get_new_name(tmp_template_name)`
and
`tmp_template_name = self.scope.get_new_name(a.name)`

Is that the former will create a name like `a.name_sdfghj` and sometimes `a.name_sdfghj_00001` while the latter will create a name like `a.name_00001`. Neither will create collisions and neither will return `a.name`. When the name already exists (as is necessarily the case here) `get_new_name` already adds a suffix. The only difference as compared to your implementation is that the suffix is not random (we wanted to avoid randomness in our code so the generated code is reproduceable)

Reply:
Ok, makes sense.

> I wanted to make the code predictable for the user even in the creation of temporary variables

You mean so there are no missed numbers? Presumably the same names should be generated every time no matter the method

The name will be printed in Python code but I agree that we don't care about this. Ok for:
`tmp_template_name = self.scope.get_new_name(tmp_template_name)`

Reply:
done in [d37fc58](https://github.com/pyccel/pyccel/pull/1728/commits/d37fc58755d0470cc40aee6275d985c2cea3a53a)